pos-print/src/main/java/com/clouway/pos/print/core/BackgroundReceiptPrintWorker.kt

```diff
@@ -0,0 +1,78 @@
+package com.clouway.pos.print.core
+
+import com.clouway.pos.print.printer.Status
+import com.google.common.util.concurrent.AbstractExecutionThreadService
+import com.google.inject.Inject
+import org.slf4j.LoggerFactory
+import java.io.IOException
+
+/**
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+class BackgroundReceiptPrintWorker @Inject constructor(private var repository: ReceiptRepository,
+                                                       private var factory: PrinterFactory,
+                                                       private var notifier: PrintNotifier,
+                                                       private var queue: PrintQueue)
+  : ReceiptPrintWorker, AbstractExecutionThreadService() {
+
+  override fun run() {
+    while (isRunning){
+      printReceipts()
+    }
+  }
+
+  private val logger = LoggerFactory.getLogger(ReceiptPrintWorker::class.java)
+
+  override fun shutDown() {
+    logger.info("Stopping background printing service")
+
+    super.shutDown()
+  }
+  override fun printReceipts() {
+    while (queue.hasNext()) {
+      val receiptWithSource = queue.next()
+      println(receiptWithSource)
+      val receipt = receiptWithSource.receipt
+
+      var printer: ReceiptPrinter? = null
+
+      try {
+        printer = factory.getPrinter(receiptWithSource.sourceIp)
+
+        val printResponse = if (receiptWithSource.isFiscal) printer.printFiscalReceipt(receipt)
+        else printer.printReceipt(receipt)
+
+        if (printResponse.warnings.contains(Status.FISCAL_RECEIPT_IS_OPEN)
+          || printResponse.warnings.contains(Status.NON_FISCAL_RECEIPT_IS_OPEN)) {
+          logger.info("Receipt printing accepted")
+          repository.finishPrinting(receipt.receiptId)
```

Comment:
I don't think the repository should be responsible for the notification, as the listener handles a PrintReceiptResponse which the repository otherwise doesn't have to know about.

Reply:
You are partially right but:

Here you have 4 dependencies:
```kotlin
class BackgroundReceiptPrintWorker @Inject constructor(private var repository: ReceiptRepository,
                                                       private var factory: PrinterFactory,
                                                       private var notifier: PrintNotifier,
                                                       private var queue: PrintQueue)
```
which causes tests to become bloating with a lot of code due the number of dependencies. This is a smell and need to be considered split of them.

Reply:
Should I make the repository depend on the listener interface or should I scrap it and write a hard-coded variant?

Reply:
Could you elaborate "write a hard-coded variant?"

Some snippet will be helpful.

Reply:
Currently I only have one implementation of the listener that waits for a second and prints to the console which I used for testing.

```code
class ConsolePrintingListener : PrintingListener {
  override fun notifyPrinted(printResponse: PrintReceiptResponse) {
    Thread.sleep(1000)
    println(printResponse)
  }
}
```

My question is if I should pass the repository this interface as a dependency, or if I should have something like a logger that logs when finishReceipt or failReceipt are called.

Reply:
Yes, just change the name from notifyPrinted to something like: onReceiptPrinted 

as listener is a class which receive events and not direct orders like "notifyPrinted".

Reply:
Okay, understood.